crates/svm-runtime-c-api/tests/api_tests.rs

```diff
@@ -1,23 +1,156 @@
+#![allow(unused)]
+
 extern crate svm_runtime_c_api;
 
 use svm_runtime_c_api as api;
-use svm_runtime_c_api::svm_byte_array;
 
+use std::convert::TryFrom;
 use std::ffi::c_void;
 
-use svm_abi_encoder::Encoder;
 use svm_codec::api::raw;
+use svm_ffi::{svm_byte_array, svm_env_t, svm_trap_t};
 use svm_layout::DataLayout;
-use svm_runtime::testing::WasmFile;
-use svm_types::{Address, State};
+use svm_runtime::{testing::WasmFile, vmcalls, Context};
+use svm_types::{Address, State, WasmType, WasmValue};
+
+use svm_sdk::traits::Encoder;
+use svm_sdk::ReturnData;
+
+/// We should land here when `trampoline` has been called with `host_env` containing
+/// a function index equaling to `COUNTER_MUL_FN_INDEX`
+fn counter_mul(ctx: &mut Context, args: &[WasmValue]) -> Result<Vec<WasmValue>, &'static str> {
+    assert_eq!(args.len(), 2);
+
+    let var_id = args[0].as_i32().unwrap();
+    let mul = args[1].as_i32().unwrap();
+
+    let old = vmcalls::get32(ctx, var_id);
+    let new = old * mul;
+
+    let results = vec![WasmValue::I32(new)];
+
+    Ok(results)
+}
+
+/// This struct will serve as our `host_env`.
+/// We use here integers as function identifiers to be called by the `trampoline`.
+#[derive(Debug)]
+#[repr(C)]
+struct func_index_t(u32);
+
+type Callback = fn(&mut Context, &[WasmValue]) -> Result<Vec<WasmValue>, &'static str>;
+
+const COUNTER_MUL_FN_INDEX: u32 = 123;
+
+fn wasm_trap(err: String) -> *mut svm_trap_t {
+    let trap: svm_trap_t = err.into();
+
+    // this heap-allocated memory will be released by SVM.
+    // (See: `ExternImport#wasmer_export`)
+
+    Box::into_raw(Box::new(trap))
```

Comment:
Same as previously with `results`, this is expected to be allocated by the extern client, and to be released by `svm`, hence can lead to a crash or a leak.

Reply:
That's a bit tricky since `SVM` can't know ahead-of-time how much space to allocate for `svm_trap_t`.
The only workaround I've in mind is to agree on a maximum trap size - say 1024 bytes for example.

Reply:
In `go-wasmer` the solution is:
1. Allocate the error in Go
2. Create C-compatible alias that Rust can use (via `to_wasm_trap_new`, defined in cgo preamble)
3. Call Rust's C-API and re-allocate it on the heap  (via `wasm_trap_new`)
4. Go allocation to be automatically released
5. Rust allocation to be released by Go (via `wasm_trap_delete`). I'm not sure about this step, because Rust is using the trap raw point (`trampoline` return value), so it's not clear how would it still be valid (if released by Go immediately after `trampoline` returns), or why Go even need to bother with timing its release. So i'm either missing something, or the code is still in the works.

Reply:
`svm-runtime-c-api` added a method of trap allocation. (also a method for deallocation - in general, it should be used only for tests since SVM is in charge of deallocation of traps)